### DISCUSSION TITLE
TCABootstrapApp - tool for bootstrapping a TCA app with Isowords-like structure

### DISCUSSION THREAD

**Discussion by Sajjon** [pointfreeco/swift-composable-architecture#1691] [Ideas]
Today it takes a couple of minutes and some for beginners non-intuitive steps to fully bootstrap an app with the structure [Isowords](https://github.com/pointfreeco/isowords)  uses:
* `Package.swift`, `Sources` and `Tests` folders in root, exactly like an SPM Package.
* Actual app inside `App` folder in root, that using the [dummy Package.swift "hack"](https://github.com/pointfreeco/isowords/blob/main/App/Package.swift) in order to "break recursion" when one adds the parent folder (root) as a Local Package dependency
* Then adding AppFeature.swift as "Link Binary with Libraries"
* Add AppFeature as "Target Dependency" (to trigger re-compilation of AppFeature and its dependency when launching app)
* And potentially make edits to the structure of the App, e.g. flatten the structure one level (move up `Assets.xcassets`, `ExampleApp.entitlements` (for multi-platform apps), `ExampleAppApp.swift` one level - which might require edits to the Xcode project file which references `ExampleApp.entitlements`), delete `ContentView.swift`
* Edit the `*App.swift` file with the `SwiftUI.App` to be using the `AppFeature`

Would it not be nice to have a CLI tool to do all this? Where you input just a name of the app? 

This can easily be extended with other useful functionality such as a set of common `.xcconfig` files, e.g. `dev.xcconfig`, `alpha.xcconfig`, `beta.xcconfig`, `release.xcconfig` for further decreasing the need of the horrible .xcodeproj file.

This Bootstrapped app project might even ship with TCA Xcode Templates for convenience?

Either this requires a sophisticated way of modifying a newly created Package.swift - but [that seems non-trivial to do](https://forums.swift.org/t/best-way-to-parse-package-swift-in-swift/55306) - or we could just bundle the Bootstrapped project with a Package.swift which defines an AppFeature and has TCA dependency added already. And potentially just change the `ToBeChanged` string:
```swift
let package = Package(
    name: "ToBeChanged",
```
using simple find replace?

And one could generate an Xcodeproj using [`XcodeGen`](https://github.com/yonaskolb/XcodeGen) maybe?

What do you guys think?

**Comment by Sajjon** [comment]
Just took a closer look at [Tuist](https://github.com/tuist/tuist/tree/main/projects/tuist/fixtures/app_with_spm_dependencies), a Tuist fixture will probably be able to power most of my ideas presented above.

**Reply by mbrandonw** [reply]
Hi @Sajjon, I'm glad you could find a set up that works for you, but I don't think this is something we want to ship directly in TCA. We don't want to be prescriptive in how people structure their features with the library. There are already a number of patterns flourishing in the community, and we have our own preferences, but their effectiveness greatly depends on team make up and size.

**Comment by mackoj** [comment]
I love the idea. I would prefer to not need to install Tuist or Xcodegen. Having a quick TCA Template tool is great.
